src/rgw/rgw_sal_motr.cc

```diff
@@ -2113,54 +2113,79 @@ int MotrObject::get_bucket_dir_ent(const DoutPrefixProvider *dpp, rgw_bucket_dir
       this->get_bucket()->get_info().versioning_status() == BUCKET_SUSPENDED) {
 
-    rgw_bucket_dir_entry ent_to_check;
+    //rgw_bucket_dir_entry ent_to_check;
 
+    // Check entry in the cache
     if (this->store->get_obj_meta_cache()->get(dpp, this->get_name(), bl) == 0) {
-      iter = bl.cbegin();
-      ent_to_check.decode(iter);
-      if (ent_to_check.is_current()) {
-        ent = ent_to_check;
+        iter = bl.cbegin();
+        ent.decode(iter);
+        //ent = ent_to_check;
         rc = 0;
         goto out;
-      }
     }
 
-    ldpp_dout(dpp, 20) <<__func__<< ": versioned bucket!" << dendl;
-    keys[0] = this->get_name();
-    rc = store->next_query_by_name(bucket_index_iname, keys, vals);
-    if (rc < 0) {
-      ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
-      return rc;
-    }
+    if(this->have_instance())
+    {
+      // TODO : Handle null version-id scenarios
 
-    rc = -ENOENT;
-    for (const auto& bl: vals) {
-      if (bl.length() == 0)
-        break;
+      // Cache miss.
+      rc = this->store->do_idx_op_by_name(bucket_index_iname,
+                          M0_IC_GET, this->get_key().to_str(), bl);
+      if(rc<0)
+        return rc;
 
       iter = bl.cbegin();
-      ent_to_check.decode(iter);
-      if (ent_to_check.is_current()) {
-        ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
-        ent = ent_to_check;
-        rc = 0;
-
-        this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+      ent.decode(iter);
+      //ent = ent_to_check;
+      rc = 0;
+      // Put into the cache
+      this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+      goto out;
 
-        break;
-      }
     }
-  } else {
-    if (this->store->get_obj_meta_cache()->get(dpp, this->get_key().to_str(), bl)) {
+    else
+    {  // Version-id instance is empty
+       // Cache miss.
+        keys[0] = this->get_name();
+
+        rc = store->next_query_by_name(bucket_index_iname, keys, vals, this->get_name());
+        if (rc < 0) {
+          ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
+          return rc;
+        }
+
+        rc = -ENOENT;
+
+        for (const auto& bl: vals) {
+          if (bl.length() == 0)
+            break;
+
+          iter = bl.cbegin();
+          ent.decode(iter);
+
+          // check the current object version
+          if (ent.is_current()) {
+            ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
+            //ent = ent_to_check;
+            rc = 0;
+            this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+            break;
+          }
+        }
+     }
+    } else {
+    if (this->store->get_obj_meta_cache()->get(dpp, this->get_name(), bl)) {
       ldpp_dout(dpp, 20) <<__func__<< ": non-versioned bucket!" << dendl;
+
+      // TODO : Handle null version-id scenarios
```

Comment:
I guess null version is handled after making using `this->get_name()` at line 2188. If true then the comment can be removed.

Reply:
Yes, after making use of `this->get_name()` we can retrieve the null object version contents. But in the response, version-id field itself can be missing. To handle this null cases, some changes are required in PUT object. 
The changes will be covered in another PR - https://github.com/Seagate/cortx-rgw/pull/200